text/0000-fix-error.md

```diff
@@ -0,0 +1,246 @@
+- Feature Name: fix_error
+- Start Date: 2018-07-18
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Change the `std::error::Error` trait to improve its usability. Introduce a
+backtrace module to the standard library to provide a standard interface for
+backtraces.
+
+# Motivation
+[motivation]: #motivation
+
+The `Error` trait has long been known to be flawed in several respects. In
+particular:
+
+1. The required `description` method is limited, usually, to returning static
+strings, and has little utility that isn't adequately served by the required
+`Display` impl for the error type.
+2. The signature of the `cause` method does not allow the user to downcast the
+cause type, limiting the utility of that method unnecessarily.
+3. It provides no standard API for errors that contain backtraces (as some
+users' errors do) to expose their backtraces to end users.
+
+We propose to fix this by deprecating the existing methods of `Error` and adding
+two new, provided methods. As a result, the undeprecated portion of the `Error`
+trait would look like this:
+
+```rust
+trait Error: Display + Debug {
+    fn backtrace(&self) -> Option<&Backtrace> {
+        None
+    }
+
+    fn source(&self) -> Option<&dyn Error + 'static> {
+        None
+    }
+}
+```
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## The new API of the Error trait
+
+The new API provides three main components:
+
+1. The Display and Debug impls, for printing error messages. Ideally, the
+   Display API would be focused on *end user* messages, whereas the Debug impl
+   would contain information relevant to the programmer.
+2. The backtrace method. If the Error contains a backtrace, it should be exposed
+   through this method. Errors are not required to contain a backtrace, and are
+   not expected to.
+3. The `source` method. This returns another Error type, which is the underlying
+   source of this error. If this error has no underlying source (that is, it is
+   the "root source" of the error), this method should return none.
+
+## The backtrace API
+
+This RFC adds a new `backtrace` module to std, with one type, with this API:
+
+```rust
+pub struct Backtrace {
+    // ...
+}
+
+impl Backtrace {
+    pub fn new() -> Backtrace {
+        // ...
+    }
+}
+
+impl Display for Backtrace {
+    // ...
+}
+
+impl Debug for Backtrace {
+    // ...
+}
+```
+
+This minimal initial API is just intended for printing backtraces for end users.
+In time, this may grow the ability to visit individual frames of the backtrace.
+
+### Backtrace controlling environment variables
+
+Today, the `RUST_BACKTRACE` controls backtraces generated by panics. After this
+RFC, it also controls backtraces generated in the standard library: no backtrace
+will be generated when calling `Backtrace::new` unless this variable is set.
```

Comment:
It seems kind of weird for it to be impossible to get a backtrace at all if the right environment variable isn't set. It seems reasonable to have a constructor that returns an `Option<Backtrace>` by checking the environment, but I think `Backtrace::new` should just unconditionally produce a backtrace.

Reply:
I agree. I effectively have to force this envvar on all the time now on startup which seems like a weird way to control the system.

Reply:
@withoutboats 
> the behavior of a Backtrace is based on whether or not that environment variable is set.

Personally, I feel like that policy should be decoupled from being able to get a `Backtrace`.  We should have an aligned default policy but we should make it possible for people to create policies to suit their needs.  For example, this RFC is bringing up a change in policy:

From RFC
> Two additional variables will be added: RUST_PANIC_BACKTRACE and RUST_STD_BACKTRACE: these will independently override the behavior of RUST_BACKTRACE for backtraces generated for panics and from the std API.

How will `Backtrace` know its calling context to use these two variables?  

@withoutboats 
>  If it isn't, its a null type that doesn't collect any backtrace information and prints nothing when displayed.

And in `Error`s case, is there a reason to use null objects rather than `Option` like the API?